hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCreateRequest.java

```diff
@@ -0,0 +1,375 @@
+    .KeyArgs;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos
+    .OMRequest;
+import org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos
+    .OMResponse;
+import org.apache.hadoop.ozone.security.acl.IAccessAuthorizer;
+import org.apache.hadoop.ozone.security.acl.OzoneObj;
+import org.apache.hadoop.util.Time;
+import org.apache.hadoop.utils.UniqueId;
+import org.apache.hadoop.utils.db.cache.CacheKey;
+import org.apache.hadoop.utils.db.cache.CacheValue;
+
+/**
+ * Handles CreateKey request.
+ */
+
+public class OMKeyCreateRequest extends OMClientRequest
+    implements OMKeyRequest {
+  private static final Logger LOG =
+      LoggerFactory.getLogger(OMKeyCreateRequest.class);
+
+  public OMKeyCreateRequest(OMRequest omRequest) {
+    super(omRequest);
+  }
+
+  @Override
+  public OMRequest preExecute(OzoneManager ozoneManager) throws IOException {
+    CreateKeyRequest createKeyRequest = getOmRequest().getCreateKeyRequest();
+    Preconditions.checkNotNull(createKeyRequest);
+
+    KeyArgs keyArgs = createKeyRequest.getKeyArgs();
+
+    // We cannot allocate block for multipart upload part when
+    // createMultipartKey is called, as we will not know type and factor with
+    // which initiateMultipartUpload has started for this key. When
+    // allocateBlock call happen's we shall know type and factor, as we set
+    // the type and factor read from multipart table, and set the KeyInfo in
+    // validateAndUpdateCache and return to the client. TODO: See if we can fix
+    //  this.
+
+    CreateKeyRequest.Builder newCreateKeyRequest = null;
+    KeyArgs.Builder newKeyArgs = null;
+    if (!keyArgs.getIsMultipartKey()) {
+
+      long scmBlockSize = ozoneManager.getScmBlockSize();
+
+      // NOTE size of a key is not a hard limit on anything, it is a value that
+      // client should expect, in terms of current size of key. If client sets
+      // a value, then this value is used, otherwise, we allocate a single
+      // block which is the current size, if read by the client.
+      final long requestedSize = keyArgs.getDataSize() > 0 ?
+          keyArgs.getDataSize() : scmBlockSize;
+
+      boolean useRatis = ozoneManager.isUseRatis();
+
+      HddsProtos.ReplicationFactor factor = keyArgs.getFactor();
+      if (factor == null) {
+        factor = useRatis ? HddsProtos.ReplicationFactor.THREE :
+            HddsProtos.ReplicationFactor.ONE;
+      }
+
+      HddsProtos.ReplicationType type = keyArgs.getType();
+      if (type == null) {
+        type = useRatis ? HddsProtos.ReplicationType.RATIS :
+            HddsProtos.ReplicationType.STAND_ALONE;
+      }
+
+      // TODO: Here we are allocating block with out any check for
+      //  bucket/key/volume or not and also with out any authorization checks.
+      //  As for a client for the first time this can be executed on any OM,
+      //  till leader is identified.
+
```

Comment:
As discussed offline, we will add a isLeader check before calling preExecute. This will help in reducing the chance of allocating blocks on follower OM's.